File: python/tests/test_m83_rans.py
"""M83 — Clean-room rANS entropy codec.

Round-trip + canonical-vector + malformed-input + throughput tests
for the order-0 and order-1 rANS codec implemented in
``ttio.codecs.rans``.  The canonical vector fixtures in
``tests/fixtures/codecs/rans_<x>_o<n>.bin`` are the cross-language
conformance contract — the ObjC and Java ports must produce
byte-exact identical outputs for the same inputs.

Spec of record: ``HANDOFF.md`` §1, §2, §5, §6, §7.1.
"""
from __future__ import annotations

import hashlib
import os
import time
from pathlib import Path

import pytest

from ttio.codecs.rans import decode, encode

FIXTURE_DIR = Path(__file__).parent / "fixtures" / "codecs"


# ── Canonical test vectors (HANDOFF.md §6.1) ────────────────────────


def _vector_a() -> bytes:
    """Uniform-random-ish 256 bytes (SHA-256 digest repeated 8x)."""
    return hashlib.sha256(b"ttio-rans-test-vector-a").digest() * 8


def _vector_b() -> bytes:
    """Heavily skewed 1024 bytes — 78% 0x00, 10% 0x01, 8% 0x02, 4% 0x03."""
    return bytes([0] * 800 + [1] * 100 + [2] * 80 + [3] * 44)


def _vector_c() -> bytes:
    """Perfectly cyclic 0,1,2,3,0,1,2,3,... (512 bytes) — order-1 ideal."""
    return bytes([i % 4 for i in range(512)])


# ── Round-trip tests (1–7) ──────────────────────────────────────────


def test_01_roundtrip_order0_random_1mb() -> None:
    """Order-0 round-trip on 1 MiB of os.urandom — byte-exact."""
    data = os.urandom(1 << 20)
    enc = encode(data, 0)
    dec = decode(enc)
    assert dec == data
    # Random data is incompressible — encoded size should be roughly
    # input size (within a small constant overhead).
    assert len(enc) >= len(data)


def test_02_roundtrip_order1_random_1mb() -> None:
    """Order-1 round-trip on 1 MiB of os.urandom — byte-exact."""
    data = os.urandom(1 << 20)
    enc = encode(data, 1)
    dec = decode(enc)
    assert dec == data


def test_03_roundtrip_order0_biased_1mb() -> None:
    """Order-0 round-trip on 1 MiB biased data; compressed < 50 % of input."""
    n = 1 << 20
    counts = (int(n * 0.90), int(n * 0.05), int(n * 0.03))
    payload = (
        bytes([0x00]) * counts[0]
        + bytes([0x01]) * counts[1]
        + bytes([0x02]) * counts[2]
    )
    payload += bytes([0x03]) * (n - len(payload))
    assert len(payload) == n
    enc = encode(payload, 0)
    dec = decode(enc)
    assert dec == payload
    assert len(enc) < n // 2, f"biased order-0 enc={len(enc)} >= n/2"


def test_04_roundtrip_order1_biased_le_order0() -> None:
    """Order-1 biased round-trip; compressed ≤ order-0 compressed."""
    n = 1 << 20
    counts = (int(n * 0.90), int(n * 0.05), int(n * 0.03))
    payload = (
        bytes([0x00]) * counts[0]
        + bytes([0x01]) * counts[1]
        + bytes([0x02]) * counts[2]
    )
    payload += bytes([0x03]) * (n - len(payload))
    enc0 = encode(payload, 0)
    enc1 = encode(payload, 1)
    assert decode(enc1) == payload
    assert len(enc1) <= len(enc0)


def test_05_roundtrip_all_identical_1mb() -> None:
    """1 MiB of 0x41 — compressed size < 10 KiB, round-trip exact."""
    data = b"\x41" * (1 << 20)
    enc = encode(data, 0)
    assert decode(enc) == data
    assert len(enc) < 10 * 1024, f"all-identical enc={len(enc)} >= 10KiB"


def test_06_roundtrip_single_byte() -> None:
    """Single-byte input round-trips exactly at both orders."""
    for order in (0, 1):
        enc = encode(b"\x42", order)
        assert decode(enc) == b"\x42"


def test_07_roundtrip_empty() -> None:
    """Empty input round-trips at both orders."""
    for order in (0, 1):
        enc = encode(b"", order)
        assert decode(enc) == b""


# ── Canonical vector fixtures (8–12) ────────────────────────────────


def _load_fixture(name: str) -> bytes:
    path = FIXTURE_DIR / name
    return path.read_bytes()


def test_08_canonical_vector_a_order0() -> None:
    """Vector A order-0 encodes byte-exactly to the committed fixture."""
    data = _vector_a()
    enc = encode(data, 0)
    assert enc == _load_fixture("rans_a_o0.bin")
    assert decode(enc) == data


def test_09_canonical_vector_a_order1() -> None:
    """Vector A order-1 encodes byte-exactly to the committed fixture."""
    data = _vector_a()
    enc = encode(data, 1)
    assert enc == _load_fixture("rans_a_o1.bin")
    assert decode(enc) == data


def test_10_canonical_vector_b_order0() -> None:
    """Vector B order-0: byte-exact match + payload < 300 bytes.

    The HANDOFF spec states "compressed size < 300 bytes" for the
    biased 1024-byte vector B.  With our wire format the order-0
    frequency table alone is 1024 bytes, so the < 300 byte budget
    is checked against the rANS payload (header + freq table is
    fixed overhead).
    """
    data = _vector_b()
    enc = encode(data, 0)
    assert enc == _load_fixture("rans_b_o0.bin")
    assert decode(enc) == data
    payload_len = int.from_bytes(enc[5:9], "big")
    assert payload_len < 300, f"vec B order-0 payload={payload_len} bytes"


def test_11_canonical_vector_b_order1() -> None:
    """Vector B order-1 encodes byte-exactly to the committed fixture."""
    data = _vector_b()
    enc = encode(data, 1)
    assert enc == _load_fixture("rans_b_o1.bin")
    assert decode(enc) == data


def test_12_canonical_vector_c_order1_smaller_than_order0() -> None:
    """Vector C: order-1 compressed size < order-0 compressed size."""
    data = _vector_c()
    enc0 = encode(data, 0)
    enc1 = encode(data, 1)
    assert enc0 == _load_fixture("rans_c_o0.bin")
    assert enc1 == _load_fixture("rans_c_o1.bin")
    assert decode(enc0) == data
    assert decode(enc1) == data
    assert len(enc1) < len(enc0), (
        f"order-1 ({len(enc1)}) should beat order-0 ({len(enc0)}) "
        "on perfectly cyclic data"
    )


# ── Malformed input (13) ────────────────────────────────────────────


def test_13_decode_malformed() -> None:
    """Truncated / bad-order / wrong-length inputs raise ValueError."""
    good = encode(b"hello world" * 100, 0)

    # Empty input.
    with pytest.raises(ValueError):
        decode(b"")
    # Shorter than header.
    with pytest.raises(ValueError):
        decode(b"\x00\x00\x00")
    # Bad order byte.
    bad_order = bytearray(good)
    bad_order[0] = 0x05
    with pytest.raises(ValueError):
        decode(bytes(bad_order))
    # Truncated payload (drop tail bytes).
    with pytest.raises(ValueError):
        decode(good[:-4])
    # Truncated freq table.
    with pytest.raises(ValueError):
        decode(good[:50])
    # Header lies about payload length — actual stream too short.
    bad_len = bytearray(good)
    declared = int.from_bytes(bad_len[5:9], "big")
    bad_len[5:9] = (declared + 16).to_bytes(4, "big")
    with pytest.raises(ValueError):
        decode(bytes(bad_len))
    # Order-1 with a context table that sums to something other than M.
    enc1 = bytearray(encode(b"abcabcabc", 1))
    # Find the first non-empty row and corrupt one of its frequencies.
    off = 9
    n = len(enc1)
    fixed = False
    for ctx in range(256):
        n_nonzero = int.from_bytes(enc1[off : off + 2], "big")
        off += 2
        if n_nonzero > 0:
            # Add 1 to the first symbol's frequency (breaks sum == M).
            f = int.from_bytes(enc1[off + 1 : off + 3], "big")
            enc1[off + 1 : off + 3] = (f + 1).to_bytes(2, "big")
            fixed = True
            break
        if off >= n:
            break
        off += n_nonzero * 3
    assert fixed, "test setup: could not find row to corrupt"
    with pytest.raises(ValueError):
        decode(bytes(enc1))

    # Type errors.
    with pytest.raises(TypeError):
        decode("not-bytes")  # type: ignore[arg-type]
    with pytest.raises(ValueError):
        encode(b"x", order=2)


# ── Throughput (14) ─────────────────────────────────────────────────


def test_14_throughput_order0_10mb() -> None:
    """Encode + decode 10 MiB order-0; print MB/s; relaxed thresholds.

    Pure-Python rANS is intentionally slow (no Cython / no numpy);
    targets are 2 MB/s encode, 5 MB/s decode per Binding Decision §82.
    """
    n = 10 * (1 << 20)  # 10 MiB
    data = os.urandom(n)

    t0 = time.perf_counter()
    enc = encode(data, 0)
    enc_dt = time.perf_counter() - t0

    t1 = time.perf_counter()
    dec = decode(enc)
    dec_dt = time.perf_counter() - t1

    enc_mb_s = (n / (1 << 20)) / enc_dt
    dec_mb_s = (n / (1 << 20)) / dec_dt
    print(
        f"\n  M83 throughput (10 MiB, order-0, pure Python): "
        f"encode {enc_mb_s:.2f} MB/s, decode {dec_mb_s:.2f} MB/s"
    )

    assert dec == data
    assert enc_mb_s >= 2.0, f"encode {enc_mb_s:.2f} MB/s < 2 MB/s"
    assert dec_mb_s >= 5.0, f"decode {dec_mb_s:.2f} MB/s < 5 MB/s"

File: objc/Tests/TestM83Rans.m
// TestM83Rans.m — v0.13 M83.
//
// Objective-C normative tests for the clean-room rANS entropy codec.
// Mirrors python/tests/test_m83_rans.py and locks the cross-language
// wire format via byte-exact fixture comparison.
//
// SPDX-License-Identifier: LGPL-3.0-or-later

#import <Foundation/Foundation.h>
#import "Testing.h"
#import "Codecs/TTIORans.h"

#include <openssl/sha.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

// ── Fixture loader (walk upward from CWD looking for objc/Tests/Fixtures)

static NSString *fixtureDir(void)
{
    NSFileManager *fm = [NSFileManager defaultManager];
    NSString *here = [fm currentDirectoryPath];
    for (int up = 0; up < 6; up++) {
        NSString *candidate = [[here
                stringByAppendingPathComponent:@"Tests"]
                stringByAppendingPathComponent:@"Fixtures"];
        BOOL isDir = NO;
        if ([fm fileExistsAtPath:candidate isDirectory:&isDir] && isDir) {
            // confirm it has at least one rans_*.bin so we don't pick a
            // wrong Tests/Fixtures higher up.
            NSString *probe = [candidate stringByAppendingPathComponent:@"rans_a_o0.bin"];
            if ([fm fileExistsAtPath:probe]) return candidate;
        }
        // Also try a sibling objc/Tests/Fixtures (when CWD is repo root).
        NSString *candidate2 = [[[here
                stringByAppendingPathComponent:@"objc"]
                stringByAppendingPathComponent:@"Tests"]
                stringByAppendingPathComponent:@"Fixtures"];
        if ([fm fileExistsAtPath:candidate2 isDirectory:&isDir] && isDir) {
            NSString *probe = [candidate2 stringByAppendingPathComponent:@"rans_a_o0.bin"];
            if ([fm fileExistsAtPath:probe]) return candidate2;
        }
        here = [here stringByDeletingLastPathComponent];
        if ([here isEqualToString:@"/"] || here.length == 0) break;
    }
    return nil;
}

static NSData *loadFixture(NSString *name)
{
    NSString *dir = fixtureDir();
    if (!dir) return nil;
    NSString *path = [dir stringByAppendingPathComponent:name];
    return [NSData dataWithContentsOfFile:path];
}

// ── Canonical vector builders (HANDOFF.md §6.1) ────────────────────

static NSData *vectorA(void)
{
    // SHA-256 of "ttio-rans-test-vector-a", repeated 8 times = 256 bytes.
    const char salt[] = "ttio-rans-test-vector-a";
    uint8_t digest[32];
    SHA256((const uint8_t *)salt, sizeof(salt) - 1, digest);
    NSMutableData *out = [NSMutableData dataWithCapacity:256];
    for (int i = 0; i < 8; i++) [out appendBytes:digest length:32];
    return out;
}

static NSData *vectorB(void)
{
    // 800 × 0x00 + 100 × 0x01 + 80 × 0x02 + 44 × 0x03 = 1024 bytes.
    NSMutableData *out = [NSMutableData dataWithLength:1024];
    uint8_t *p = (uint8_t *)out.mutableBytes;
    memset(p, 0, 800);
    memset(p + 800, 1, 100);
    memset(p + 900, 2, 80);
    memset(p + 980, 3, 44);
    return out;
}

static NSData *vectorC(void)
{
    // 0,1,2,3,0,1,2,3,... × 128 = 512 bytes.
    NSMutableData *out = [NSMutableData dataWithLength:512];
    uint8_t *p = (uint8_t *)out.mutableBytes;
    for (int i = 0; i < 512; i++) p[i] = (uint8_t)(i % 4);
    return out;
}

// ── Helpers ────────────────────────────────────────────────────────

static NSData *randomData(NSUInteger n, unsigned int seed)
{
    NSMutableData *out = [NSMutableData dataWithLength:n];
    uint8_t *p = (uint8_t *)out.mutableBytes;
    // Linear congruential filler — deterministic per seed but not
    // patterned in any way the codec is sensitive to.  The codec only
    // sees byte values, so this is plenty random for round-trip tests.
    uint32_t state = seed ? seed : 0xDEADBEEFu;
    for (NSUInteger i = 0; i < n; i++) {
        state = state * 1664525u + 1013904223u;
        p[i] = (uint8_t)((state >> 16) & 0xFF);
    }
    return out;
}

static NSData *biasedData(NSUInteger n)
{
    // 90% 0x00, 5% 0x01, 3% 0x02, 2% 0x03 — block layout, same shape
    // as the Python test (counts are deterministic given n).
    NSUInteger c0 = (NSUInteger)((double)n * 0.90);
    NSUInteger c1 = (NSUInteger)((double)n * 0.05);
    NSUInteger c2 = (NSUInteger)((double)n * 0.03);
    NSUInteger c3 = (n > c0 + c1 + c2) ? (n - c0 - c1 - c2) : 0;
    NSMutableData *out = [NSMutableData dataWithLength:n];
    uint8_t *p = (uint8_t *)out.mutableBytes;
    NSUInteger pos = 0;
    memset(p + pos, 0, c0); pos += c0;
    memset(p + pos, 1, c1); pos += c1;
    memset(p + pos, 2, c2); pos += c2;
    memset(p + pos, 3, c3);
    return out;
}

static double monoSeconds(void)
{
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec + (double)ts.tv_nsec * 1e-9;
}

// ── Tests #1–#6: round-trip ────────────────────────────────────────

static void testRoundTripRandomOrder0(void)
{
    NSData *data = randomData(1u << 20, 1);
    NSData *enc = TTIORansEncode(data, 0);
    PASS(enc != nil, "M83: order-0 1MB random encode succeeds");
    PASS(enc.length > 0,
         "M83: order-0 1MB random encode emits bytes");
    NSError *err = nil;
    NSData *dec = TTIORansDecode(enc, &err);
    PASS(dec != nil, "M83: order-0 1MB random decode succeeds");
    PASS([dec isEqualToData:data],
         "M83: order-0 1MB random round-trips byte-exact");
}

static void testRoundTripRandomOrder1(void)
{
    NSData *data = randomData(1u << 20, 2);
    NSData *enc = TTIORansEncode(data, 1);
    PASS(enc != nil, "M83: order-1 1MB random encode succeeds");
    NSError *err = nil;
    NSData *dec = TTIORansDecode(enc, &err);
    PASS(dec != nil, "M83: order-1 1MB random decode succeeds");
    PASS([dec isEqualToData:data],
         "M83: order-1 1MB random round-trips byte-exact");
}

static void testRoundTripBiasedOrder0(void)
{
    NSData *data = biasedData(1u << 20);
    NSData *enc = TTIORansEncode(data, 0);
    PASS(enc != nil, "M83: order-0 1MB biased encode succeeds");
    PASS(enc.length < (1u << 19),
         "M83: order-0 1MB biased compresses to < 0.5 MB (got %lu)",
         (unsigned long)enc.length);
    NSError *err = nil;
    NSData *dec = TTIORansDecode(enc, &err);
    PASS([dec isEqualToData:data],
         "M83: order-0 1MB biased round-trips byte-exact");
}

static void testRoundTripAllIdentical(void)
{
    NSMutableData *data = [NSMutableData dataWithLength:1u << 20];
    memset(data.mutableBytes, 0x41, data.length);
    NSData *enc = TTIORansEncode(data, 0);
    PASS(enc != nil, "M83: all-identical encode succeeds");
    PASS(enc.length < 10 * 1024,
         "M83: 1 MB of 0x41 compresses to < 10 KB (got %lu)",
         (unsigned long)enc.length);
    NSError *err = nil;
    NSData *dec = TTIORansDecode(enc, &err);
    PASS([dec isEqualToData:data],
         "M83: all-identical round-trips byte-exact");
}

static void testRoundTripEmpty(void)
{
    for (int order = 0; order <= 1; order++) {
        NSData *enc = TTIORansEncode([NSData data], order);
        PASS(enc != nil && enc.length >= 9,
             "M83: empty order-%d encode produces >= 9 byte stream", order);
        NSError *err = nil;
        NSData *dec = TTIORansDecode(enc, &err);
        PASS(dec != nil && dec.length == 0,
             "M83: empty order-%d round-trips to empty", order);
    }
}

static void testRoundTripSingleByte(void)
{
    uint8_t b = 0x42;
    NSData *data = [NSData dataWithBytes:&b length:1];
    for (int order = 0; order <= 1; order++) {
        NSData *enc = TTIORansEncode(data, order);
        PASS(enc != nil, "M83: single-byte order-%d encode succeeds", order);
        NSError *err = nil;
        NSData *dec = TTIORansDecode(enc, &err);
        PASS([dec isEqualToData:data],
             "M83: single-byte order-%d round-trips byte-exact", order);
    }
}

// ── Tests #7–#10: canonical-vector byte-exact conformance ──────────

static void compareDataDumpHexOnFail(NSData *got, NSData *want, const char *label)
{
    if ([got isEqualToData:want]) return;
    NSUInteger min = (got.length < want.length) ? got.length : want.length;
    NSUInteger firstDiff = (NSUInteger)-1;
    const uint8_t *g = got.bytes;
    const uint8_t *w = want.bytes;
    for (NSUInteger i = 0; i < min; i++) {
        if (g[i] != w[i]) { firstDiff = i; break; }
    }
    fprintf(stderr,
            "  %s mismatch: got=%lu want=%lu first diff @ %ld\n",
            label, (unsigned long)got.length, (unsigned long)want.length,
            (long)firstDiff);
    if (firstDiff != (NSUInteger)-1) {
        NSUInteger lo = (firstDiff > 8) ? firstDiff - 8 : 0;
        NSUInteger hi = (firstDiff + 8 < min) ? firstDiff + 8 : min;
        fprintf(stderr, "    got: ");
        for (NSUInteger i = lo; i < hi; i++)
            fprintf(stderr, "%02x ", g[i]);
        fprintf(stderr, "\n   want: ");
        for (NSUInteger i = lo; i < hi; i++)
            fprintf(stderr, "%02x ", w[i]);
        fprintf(stderr, "\n");
    }
}

static void testCanonicalVectorAOrder0(void)
{
    NSData *fixture = loadFixture(@"rans_a_o0.bin");
    PASS(fixture != nil, "M83: rans_a_o0.bin fixture loads");
    NSData *data = vectorA();
    NSData *enc = TTIORansEncode(data, 0);
    compareDataDumpHexOnFail(enc, fixture, "vector A o0");
    PASS([enc isEqualToData:fixture],
         "M83: vector A order-0 byte-exact match against Python fixture");
    NSError *err = nil;
    NSData *dec = TTIORansDecode(fixture, &err);
    PASS([dec isEqualToData:data],
         "M83: vector A order-0 fixture decodes back to vector A");
}

static void testCanonicalVectorBOrder0(void)
{
    NSData *fixture = loadFixture(@"rans_b_o0.bin");
    PASS(fixture != nil, "M83: rans_b_o0.bin fixture loads");
    NSData *data = vectorB();
    NSData *enc = TTIORansEncode(data, 0);
    compareDataDumpHexOnFail(enc, fixture, "vector B o0");
    PASS([enc isEqualToData:fixture],
         "M83: vector B order-0 byte-exact match against Python fixture");
    // Spec: payload portion (after 9-byte header + 1024-byte freq table)
    // is < 300 bytes for the heavily skewed input.
    const uint8_t *p = enc.bytes;
    uint32_t payload_len = ((uint32_t)p[5] << 24) | ((uint32_t)p[6] << 16) |
                           ((uint32_t)p[7] <<  8) | ((uint32_t)p[8]);
    PASS(payload_len < 300,
         "M83: vector B order-0 payload < 300 bytes (got %u)", payload_len);
}

static void testCanonicalVectorCOrder0(void)
{
    NSData *fixture = loadFixture(@"rans_c_o0.bin");
    PASS(fixture != nil, "M83: rans_c_o0.bin fixture loads");
    NSData *data = vectorC();
    NSData *enc = TTIORansEncode(data, 0);
    compareDataDumpHexOnFail(enc, fixture, "vector C o0");
    PASS([enc isEqualToData:fixture],
         "M83: vector C order-0 byte-exact match against Python fixture");
}

static void testCanonicalVectorCOrder1(void)
{
    NSData *fixture0 = loadFixture(@"rans_c_o0.bin");
    NSData *fixture1 = loadFixture(@"rans_c_o1.bin");
    PASS(fixture1 != nil, "M83: rans_c_o1.bin fixture loads");
    NSData *data = vectorC();
    NSData *enc1 = TTIORansEncode(data, 1);
    compareDataDumpHexOnFail(enc1, fixture1, "vector C o1");
    PASS([enc1 isEqualToData:fixture1],
         "M83: vector C order-1 byte-exact match against Python fixture");
    PASS(fixture1.length < fixture0.length,
         "M83: vector C order-1 (%lu) beats order-0 (%lu) on cyclic data",
         (unsigned long)fixture1.length, (unsigned long)fixture0.length);
    NSError *err = nil;
    NSData *dec = TTIORansDecode(fixture1, &err);
    PASS([dec isEqualToData:data],
         "M83: vector C order-1 fixture decodes back to vector C");
}

// ── Test #11: malformed input handling ─────────────────────────────

static void testDecodeMalformed(void)
{
    NSError *err = nil;

    // Empty input.
    err = nil;
    PASS(TTIORansDecode([NSData data], &err) == nil && err != nil,
         "M83: empty input → nil + error (no crash)");

    // Shorter than header.
    err = nil;
    uint8_t tiny[3] = {0, 0, 0};
    NSData *t1 = [NSData dataWithBytes:tiny length:3];
    PASS(TTIORansDecode(t1, &err) == nil && err != nil,
         "M83: 3-byte input → nil + error");

    // Bogus order byte.
    NSData *good = TTIORansEncode([@"hello world" dataUsingEncoding:NSASCIIStringEncoding], 0);
    NSMutableData *badOrder = [good mutableCopy];
    ((uint8_t *)badOrder.mutableBytes)[0] = 0x05;
    err = nil;
    PASS(TTIORansDecode(badOrder, &err) == nil && err != nil,
         "M83: bogus order byte → nil + error");

    // Truncated payload.
    NSData *truncated = [good subdataWithRange:NSMakeRange(0, good.length - 4)];
    err = nil;
    PASS(TTIORansDecode(truncated, &err) == nil && err != nil,
         "M83: truncated payload → nil + error");

    // Truncated freq table (cut into the freq table).
    NSData *truncatedFt = [good subdataWithRange:NSMakeRange(0, 50)];
    err = nil;
    PASS(TTIORansDecode(truncatedFt, &err) == nil && err != nil,
         "M83: truncated freq table → nil + error");

    // Decoder must not crash and must produce nil — the suite passes
    // as long as we got here.
    PASS(YES, "M83: malformed-input decoding completes without crash");
}

// ── Test #12: throughput benchmark ─────────────────────────────────

static void testThroughput(void)
{
    const NSUInteger n = 4u << 20;   // 4 MiB
    NSData *data = randomData(n, 0xC0FFEE);

    double t0 = monoSeconds();
    NSData *enc = TTIORansEncode(data, 0);
    double tEnc = monoSeconds() - t0;

    double t1 = monoSeconds();
    NSError *err = nil;
    NSData *dec = TTIORansDecode(enc, &err);
    double tDec = monoSeconds() - t1;

    double mb = (double)n / (1024.0 * 1024.0);
    double encMBs = mb / tEnc;
    double decMBs = mb / tDec;
    fprintf(stderr,
            "  M83 throughput (4 MiB, order-0): "
            "encode %.1f MB/s (%.3fs), decode %.1f MB/s (%.3fs)\n",
            encMBs, tEnc, decMBs, tDec);

    PASS([dec isEqualToData:data], "M83: throughput: round-trip byte-exact");

    // Hard-fail floor at 50% of the soft target (per HANDOFF gotchas).
    PASS(encMBs >= 25.0,
         "M83: encode throughput ≥ 25 MB/s hard floor (got %.1f MB/s, soft target 50)",
         encMBs);
    PASS(decMBs >= 100.0,
         "M83: decode throughput ≥ 100 MB/s hard floor (got %.1f MB/s, soft target 200)",
         decMBs);
}

// ── Public entry point ─────────────────────────────────────────────

void testM83Rans(void);
void testM83Rans(void)
{
    testRoundTripRandomOrder0();
    testRoundTripRandomOrder1();
    testRoundTripBiasedOrder0();
    testRoundTripAllIdentical();
    testRoundTripEmpty();
    testRoundTripSingleByte();
    testCanonicalVectorAOrder0();
    testCanonicalVectorBOrder0();
    testCanonicalVectorCOrder0();
    testCanonicalVectorCOrder1();
    testDecodeMalformed();
    testThroughput();
}
